packages/eslint-plugin/src/rules/prefer-for-of.ts

```diff
@@ -0,0 +1,198 @@
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    function isSingleVariableDeclaration(
+      node: TSESTree.Node | null,
+    ): node is TSESTree.VariableDeclaration {
+      return (
+        node !== null &&
+        node.type === AST_NODE_TYPES.VariableDeclaration &&
+        node.kind !== 'const' &&
+        node.declarations.length === 1
+      );
+    }
+
+    function isLiteral(node: TSESTree.Expression, value: number): boolean {
+      return node.type === AST_NODE_TYPES.Literal && node.value === value;
+    }
+
+    function isZeroInitialized(node: TSESTree.VariableDeclarator): boolean {
+      return node.init !== null && isLiteral(node.init, 0);
+    }
+
+    function isMatchingIdentifier(
+      node: TSESTree.Expression,
+      name: string,
+    ): boolean {
+      return node.type === AST_NODE_TYPES.Identifier && node.name === name;
+    }
+
+    function isLessThanLengthExpression(
+      node: TSESTree.Node | null,
+      name: string,
+    ): TSESTree.Expression | null {
+      let arrayExpression: TSESTree.Expression | null = null;
+      if (
+        node !== null &&
+        node.type === AST_NODE_TYPES.BinaryExpression &&
+        node.operator === '<' &&
+        isMatchingIdentifier(node.left, name) &&
+        node.right.type === AST_NODE_TYPES.MemberExpression &&
+        isMatchingIdentifier(node.right.property, 'length')
+      ) {
+        arrayExpression = node.right.object;
+      }
+      return arrayExpression;
+    }
+
+    function isIncrement(node: TSESTree.Node | null, name: string): boolean {
+      if (!node) {
+        return false;
+      }
+      switch (node.type) {
+        case AST_NODE_TYPES.UpdateExpression:
+          // x++ or ++x
+          return (
+            node.operator === '++' && isMatchingIdentifier(node.argument, name)
+          );
+        case AST_NODE_TYPES.AssignmentExpression:
+          if (isMatchingIdentifier(node.left, name)) {
+            if (node.operator === '+=') {
+              // x += 1
+              return isLiteral(node.right, 1);
+            } else if (node.operator === '=') {
+              // x = x + 1 or x = 1 + x
+              const expr = node.right;
+              return (
+                expr.type === AST_NODE_TYPES.BinaryExpression &&
+                expr.operator === '+' &&
+                ((isMatchingIdentifier(expr.left, name) &&
+                  isLiteral(expr.right, 1)) ||
+                  (isLiteral(expr.left, 1) &&
+                    isMatchingIdentifier(expr.right, name)))
+              );
+            }
+          }
+      }
+      return false;
+    }
+
+    function contains(outer: TSESTree.Node, inner: TSESTree.Node): boolean {
+      return (
+        outer.range[0] <= inner.range[0] && outer.range[1] >= inner.range[1]
+      );
+    }
+
+    function isAssignee(node: TSESTree.Node): boolean {
+      const parent = node.parent;
+      if (!parent) {
+        return false;
+      }
+      return (
```

Comment:
I love the comments you've got to explain!
The only thing I would suggest is to break this up, because following the multiline boolean logic is tricky.

I know this adds a number of lines, but I think this is easier to read:
```TS
// a[i] = 1, a[i] += 1, etc.
if (
  parent.type === AST_NODE_TYPES.AssignmentExpression &&
  parent.left === node
) {
  return true;
}

// delete a[i]
if (
  parent.type === AST_NODE_TYPES.UnaryExpression &&
  parent.operator === "delete" &&
  parent.argument === node
) {
  return true;
}

// a[i]++, --a[i], etc.
// ........ etc
```